admission-webhook/webhook.go

```diff
@@ -0,0 +1,402 @@
+		logrus.Debugf("sending response: %s", responseBytes)
+
+		if _, err = responseWriter.Write(responseBytes); err != nil {
+			abortHTTPRequest(responseWriter, http.StatusInternalServerError, "error when writing response JSON %s: %v", responseBytes, err)
+		}
+	} else {
+		abortHTTPRequest(responseWriter, http.StatusInternalServerError, "error when marshalling response %v: %v", responseAdmissionReview, err)
+	}
+}
+
+// abortHTTPRequest is called for low-level HTTP errors (routing or writing the body)
+func abortHTTPRequest(responseWriter http.ResponseWriter, httpCode int, logMsg string, logArs ...interface{}) {
+	logrus.Infof(logMsg, logArs...)
+	responseWriter.WriteHeader(httpCode)
+}
+
+// httpRequestToAdmissionResponse turns a raw HTTP request into an AdmissionResponse struct.
+func (webhook *webhook) httpRequestToAdmissionResponse(request *http.Request, operation webhookOperation) *admissionv1beta1.AdmissionResponse {
+	// read the body
+	if request.Body == nil {
+		deniedAdmissionResponse(fmt.Errorf("no request body"), http.StatusBadRequest)
+	}
+	body, err := ioutil.ReadAll(request.Body)
+	if err != nil {
+		return deniedAdmissionResponse(fmt.Errorf("couldn't read request body: %v", err), http.StatusBadRequest)
+	}
+	defer request.Body.Close()
+
+	logrus.Debugf("handling %s request: %s", operation, body)
+
+	// unmarshall the request
+	admissionReview := admissionv1beta1.AdmissionReview{}
+	if err = json.Unmarshal(body, &admissionReview); err != nil {
+		return deniedAdmissionResponse(fmt.Errorf("unable to unmarshall JSON body as an admission review: %v", err), http.StatusBadRequest)
+	}
+	if admissionReview.Request == nil {
+		return deniedAdmissionResponse(fmt.Errorf("no 'Request' field in JSON body"), http.StatusBadRequest)
+	}
+
+	admissionResponse, admissionError := webhook.validateOrMutate(admissionReview.Request, operation)
+	if admissionError != nil {
+		admissionResponse = deniedAdmissionResponse(admissionError)
+	}
+
+	// return the same UID
+	admissionResponse.UID = admissionReview.Request.UID
+
+	return admissionResponse
+}
+
+// validateOrMutate is where the non-HTTP-related work happens.
+func (webhook *webhook) validateOrMutate(request *admissionv1beta1.AdmissionRequest, operation webhookOperation) (*admissionv1beta1.AdmissionResponse, *podAdmissionError) {
+	if request.Kind.Kind != "Pod" {
+		return nil, &podAdmissionError{error: fmt.Errorf("expected a Pod object, got a %v", request.Kind.Kind), code: http.StatusBadRequest}
+	}
+
+	pod, err := unmarshallPod(request.Object)
+	if err != nil {
+		return nil, err
+	}
+
+	switch request.Operation {
+	case admissionv1beta1.Create:
+		switch operation {
+		case validate:
+			return webhook.validateCreateRequest(pod, request.Namespace)
+		case mutate:
+			return webhook.mutateCreateRequest(pod)
+		default:
+			// shouldn't happen, but needed so that all paths in the function have a return value
+			panic(fmt.Errorf("unexpected webhook operation: %v", operation))
```

Comment:
The reasoning would be that since that's a switch on an enum that's hard-coded in the very same file, hitting the default case means someone added a new value to the enum without considering what how to handle it here; this IMO should trigger a very clear sign that the webhook is not at all behaving as expected, and generate a nice clear backtrace.

But then open to other suggestions if you don't like this :)